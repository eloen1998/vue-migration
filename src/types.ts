export type keysMap = {
    propsKey: string[]
    dataKey: string[]
    computedKey: string[]
}
