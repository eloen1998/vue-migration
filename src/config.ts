
export const UN_IDENTIFIED = 'UN_IDENTIFIED'