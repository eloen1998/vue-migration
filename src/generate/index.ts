export { generateComputed } from './computed'
export { generateCreated } from './created'
export { generateData } from './data'
export { generateImport } from './import'
export { generateMethod } from './method'
export { generateProps } from './props'
export { generateWatch } from './watch'
export { wrapWithComment } from './other'
