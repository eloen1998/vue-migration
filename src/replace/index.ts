
import { keysMap } from '../types'

export function replaceThis(code: string, param: keysMap) {
    return code.replace(/this\.([a-zA-Z_$][a-zA-Z0-9_$]*)/g, (subString: string, group1: string) => {
        const key = group1.trim()

        if ([...param.dataKey, ...param.computedKey].includes(key)) {
            return `${key.trim()}.value`
        }
        if (param.propsKey.includes(key)) {
            return `propsData.${key.trim()}`
        }
        if (key === '$router') {
            return 'router'
        }
        if (key === '$route') {
            return 'route'
        }

        return key.trim()
    })
}

export function replaceRefs(code: string) {
    return code.replace(/this\.\$refs([a-zA-Z_$][a-zA-Z0-9_$]*)/g, (subString: string, key: string) => {
        return key.trim()
    })
}

/**
 * 字符串的 replace 方法，第二个参数为函数时
 * 每一次匹配到内容，都会执行该函数
 * 该函数的参数
 * 第一个：匹配到的字符串
 * 第二个-第n个： 匹配到的组信息，即正则中小括号括起来的部分，有几个小括号，这里就有几个对应的参数
 * 倒数第二个参数：匹配到的字符串在原始字符串中的位置
 * 最后一个参数：原始字符串
 * 返回值： 用于替换匹配的字符串 即第一个参数
 */
