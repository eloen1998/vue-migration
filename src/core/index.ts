import { parse } from './parse'
import { generateImport, generateProps, generateData, generateComputed, generateMethod, generateWatch, wrapWithComment } from '../generate'
import { replaceThis, replaceRefs } from '../replace'

import type { ExportDefaultDeclaration, ObjectExpression } from '@babel/types'
import { generateCreated } from 'src/generate/created'
import { keysMap } from '../types'

export function transformScript(code: string) {
    const ast = parse(code)
    let result = ''

    const body = ast.program.body

    body.forEach(node => {
        if (node.type === 'ImportDeclaration') {
            result += generateImport(node)
        }

        if (node.type === 'ExportDefaultDeclaration') {
            const vueResult = formatExport(node)
            let code = replaceRefs(vueResult.code)
            code = replaceThis(code, vueResult.keysMap)
            result += code
        }
    })
    return result
}

function formatExport(node: ExportDefaultDeclaration) {
    let result = ''
    const keysMap: keysMap = {
        propsKey: [],
        dataKey: [],
        computedKey: []
    }
    if (node.declaration.type === 'ObjectExpression') {
        node.declaration.properties.forEach(node => {
            if (node.type === 'ObjectProperty') {
                if (node.key.type === 'Identifier') {
                    const { name } = node.key
                    if (name === 'props') {
                        const { code, keys } = generateProps(node.value as ObjectExpression)
                        result += code
                        keysMap.propsKey = keys
                    } else if (name === 'computed') {
                        const { code, keys } = generateComputed(node.value as ObjectExpression)
                        result += code
                        keysMap.computedKey = keys
                    } else if (name === 'filters') {
                        result += generateMethod(node.value as ObjectExpression, 'filters')
                    } else if (name === 'methods') {
                        result += generateMethod(node.value as ObjectExpression)
                    } else if (name === 'watch') {
                        result += generateWatch(node.value as ObjectExpression)
                    } else if (name === 'mixins') {
                        result += wrapWithComment(node.value as ObjectExpression, 'mixins')
                    }
                }
            }
            if (node.type === 'ObjectMethod') {
                if (node.key.type === 'Identifier') {
                    const { name } = node.key
                    if (name === 'data') {
                        const returnStatement = node.body.body.at(-1)
                        if (returnStatement && returnStatement.type === 'ReturnStatement' && returnStatement.argument?.type === 'ObjectExpression') {
                            const { code, keys } = generateData(returnStatement.argument)
                            result += code
                            keysMap.dataKey = keys
                        }
                    } else if (name === 'created') {
                        result += '/** created */\n'
                        result += generateCreated(node.body)
                    } else if (name === 'mounted') {
                        result += `onMounted(() => {\n${generateCreated(node.body)}})\n`
                    } else if (name === 'destroyed') {
                        result += `unmounted(() => {\n${generateCreated(node.body)}})\n`
                    } else if (name === 'beforeDestroy') {
                        result += `onBeforeUnmount(() => {\n${generateCreated(node.body)}})\n`
                    }
                }
            }
        })
    }
    return {
        code: result,
        keysMap
    }
}
