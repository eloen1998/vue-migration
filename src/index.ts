import { transformScript } from './core'
export default transformScript

// import { transformScript } from './core'

// const code = `
// import { isArray, isNil, isString } from 'lodash';
// export default {
//     props: {
//         value: {
//             type: String | Array,
//             default: ''
//         }
//     },
//     filters: {
//         show(val) {
//             if (isNil(val) || val === '') {
//                 return '- -';
//             }
//             return val;
//         }
//     },
//     computed: {
//         list() {
//             return isArray(this.value) ? this.value : [this.value];
//         }
//     },
//     methods: {
//         isImage(url) {
//             return isString(url) && /\.(png|jpg|jpeg|gif|bmp|webp|svg|tiff|ico)$/i.test(url);
//         }
//     }
// };`

// transformScript(code)
