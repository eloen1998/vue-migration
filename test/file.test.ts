import fs from 'fs'
import { transformScript } from '../src/core'
import { expect, test, describe } from 'vitest'

describe('整体测试', () => {

    test('语法转换', () => {

        const code = fs.readFileSync('test/input.js', 'utf-8')

        const result = transformScript(code)
        expect(result).matchSnapshot()
    })
})
