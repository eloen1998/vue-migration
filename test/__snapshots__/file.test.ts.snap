// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`整体测试 > 语法转换 1`] = `
"import { tokenQa } from './constant';
/* mixins
[tokenQa] */
const propsData = defineProps({
  prop1: {
    type: String | Array,
    default: ''
  },
  prop2: String
});
// 第一个key
const key1 = ref(0)
const key2 = ref('string')
const key3 = ref([0, 1, 2])
const key4 = ref({
  field1: '',
  field12: undefined
})
const key5 = ref([{
  value: 1,
  label: '选项1'
}])
const key6 = ref(tokenQa)
/** computed */
// 第一个computed
const key7 = computed(() => {
  return key1.value + key2.value;
})
const key8 = computed(function () {
  return key1.value + 3;
})
const key9 = computed({
  get: function () {
    return key1.value + 1;
  },
  set: function (v) {
    key1.value = v - 1;
  }
})
/** watch */
// 第一个watch
watch(watch1, (n) => {
  method2(n);
})
watch(watch2, (n,oldn) => {
  method3(n, oldn);
})
watch(watch3, (n) => {
  method2(n);
},
{
immediate: true,
deep: true})
watch(watch4, (n, oldn) => {
  method3(n, oldn);
})
/** filters */
// 第一个filter
function filter1(val) {
  return val + '1';
} 
const filter2 = function (val) {
  return val + '2';
}
/** created */
const id = router.query.id;
method1(id);
onMounted(() => {
console.log('mounted');
})
onBeforeUnmount(() => {
console.log('beforeDestroy');
})
/** methods */
/**
         * 打印id
         * @param {string} id
         */
function method1(id) {
  console.log('method1', id);
} 
// 方法2
const method2 = function () {
  console.log('method2');
}
const method3 = () => {
  // 方法3
  console.log('method3');
}
"
`;
